File: ScriptureAloneCore/Sources/ScriptureAloneCore/Keepsake/KeepsakeCrypto.swift
import Foundation
import CryptoKit
import CommonCrypto

/// Optional passphrase protection for a keepsake.
///
/// - Key: PBKDF2-HMAC-SHA256 over the passphrase (Unicode NFC, UTF-8) with a random 16-byte
///   salt and 600,000 iterations (OWASP's 2023 guidance for PBKDF2-SHA256), giving 32 bytes.
/// - Cipher: AES-256-GCM (CryptoKit) with a random 12-byte nonce. The stored payload is
///   CryptoKit's "combined" form: nonce ‖ ciphertext ‖ 16-byte tag.
/// - Associated data: the exact bytes of the outer `manifest.json`, so the salt, iteration
///   count and version beside the payload can't be altered without the open failing.
///
/// PBKDF2 is used rather than a memory-hard KDF because it ships with every Apple OS
/// (CommonCrypto) and has implementations in every language — a family member's
/// technically-minded grandchild can still open the file in fifty years.
public enum KeepsakeCrypto {
    public static let algorithm = "AES-256-GCM"
    public static let kdf = "PBKDF2-HMAC-SHA256"
    public static let defaultIterations = 600_000
    public static let saltLength = 16

    public static func randomSalt() -> Data {
        var bytes = [UInt8](repeating: 0, count: saltLength)
        var generator = SystemRandomNumberGenerator()
        for i in bytes.indices { bytes[i] = UInt8.random(in: .min ... .max, using: &generator) }
        return Data(bytes)
    }

    public static func deriveKey(passphrase: String, salt: Data, iterations: Int) throws -> SymmetricKey {
        guard iterations >= 1, iterations <= 50_000_000, !salt.isEmpty else { throw KeepsakeError.damaged("encryption settings") }
        let password = Array(passphrase.precomposedStringWithCanonicalMapping.utf8)
        guard !password.isEmpty else { throw KeepsakeError.passphraseRequired }
        var derived = [UInt8](repeating: 0, count: 32)
        let status = password.withUnsafeBufferPointer { passwordBytes in
            salt.withUnsafeBytes { saltBytes in
                passwordBytes.baseAddress!.withMemoryRebound(to: CChar.self, capacity: password.count) { passwordPointer in
                    CCKeyDerivationPBKDF(CCPBKDFAlgorithm(kCCPBKDF2),
                                         passwordPointer, password.count,
                                         saltBytes.bindMemory(to: UInt8.self).baseAddress, salt.count,
                                         CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256), UInt32(iterations),
                                         &derived, derived.count)
                }
            }
        }
        guard status == kCCSuccess else { throw KeepsakeError.damaged("key derivation failed (\(status))") }
        return SymmetricKey(data: derived)
    }

    static func seal(_ plaintext: Data, key: SymmetricKey, associatedData: Data) throws -> Data {
        let box = try AES.GCM.seal(plaintext, using: key, authenticating: associatedData)
        guard let combined = box.combined else { throw KeepsakeError.damaged("sealing failed") }
        return combined
    }

    static func open(_ combined: Data, key: SymmetricKey, associatedData: Data) throws -> Data {
        let box: AES.GCM.SealedBox
        do {
            box = try AES.GCM.SealedBox(combined: combined)
        } catch {
            throw KeepsakeError.damaged("encrypted payload")
        }
        do {
            return try AES.GCM.open(box, using: key, authenticating: associatedData)
        } catch {
            // GCM can't tell a wrong key from tampering; a wrong passphrase is by far the likelier.
            throw KeepsakeError.wrongPassphrase
        }
    }
}

File: android/app/src/main/java/com/blainemiller/scripturealone/data/keepsake/KeepsakeCrypto.kt
package com.blainemiller.scripturealone.data.keepsake

import java.security.GeneralSecurityException
import java.security.SecureRandom
import java.text.Normalizer
import javax.crypto.AEADBadTagException
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Optional passphrase protection for a keepsake, ported from `Keepsake/KeepsakeCrypto.swift`.
 *
 * - Key: PBKDF2-HMAC-SHA256 over the passphrase (Unicode NFC, UTF-8) with a random 16-byte salt and
 *   600,000 iterations, giving 32 bytes.
 * - Cipher: AES-256-GCM with a random 12-byte nonce. The stored payload is nonce ‖ ciphertext ‖
 *   16-byte tag — CryptoKit's "combined" form.
 * - Associated data: the exact bytes of the outer `manifest.json`.
 *
 * PBKDF2 is computed here over explicit UTF-8 bytes rather than through `SecretKeyFactory`, whose
 * char-to-byte conversion has differed between Android providers; the passphrase must derive the
 * same key the iPhone derives, byte for byte.
 */
object KeepsakeCrypto {
    const val ALGORITHM = "AES-256-GCM"
    const val KDF = "PBKDF2-HMAC-SHA256"
    const val DEFAULT_ITERATIONS = 600_000
    const val SALT_LENGTH = 16
    private const val NONCE_LENGTH = 12
    private const val TAG_BITS = 128
    private const val PROGRESS_STEP = 5_000

    private val random = SecureRandom()

    fun randomSalt(): ByteArray = ByteArray(SALT_LENGTH).also(random::nextBytes)

    /**
     * The passphrase's key. Deliberately slow (600,000 rounds — seconds on a phone), so callers stay
     * off the main thread; [progress] hears 0…1 as the rounds run, for a progress bar.
     */
    fun deriveKey(passphrase: String, salt: ByteArray, iterations: Int, progress: ((Float) -> Unit)? = null): ByteArray {
        if (iterations !in 1..50_000_000 || salt.isEmpty()) throw KeepsakeException.Damaged("encryption settings")
        val password = Normalizer.normalize(passphrase, Normalizer.Form.NFC).toByteArray(Charsets.UTF_8)
        if (password.isEmpty()) throw KeepsakeException.PassphraseRequired()
        return pbkdf2Sha256(password, salt, iterations, 32, progress)
    }

    /** RFC 8018 PBKDF2 with HMAC-SHA256. */
    internal fun pbkdf2Sha256(
        password: ByteArray, salt: ByteArray, iterations: Int, length: Int, progress: ((Float) -> Unit)? = null,
    ): ByteArray {
        val mac = Mac.getInstance("HmacSHA256").apply { init(SecretKeySpec(password, "HmacSHA256")) }
        val hLen = mac.macLength
        val out = ByteArray(length)
        var block = 1
        var offset = 0
        val u = ByteArray(hLen)
        val total = ((length + hLen - 1) / hLen).toLong() * iterations
        while (offset < length) {
            mac.update(salt)
            mac.update(byteArrayOf((block ushr 24).toByte(), (block ushr 16).toByte(), (block ushr 8).toByte(), block.toByte()))
            mac.doFinal(u, 0)
            val t = u.copyOf()
            for (i in 1 until iterations) {
                mac.update(u)
                mac.doFinal(u, 0)
                for (k in 0 until hLen) t[k] = (t[k].toInt() xor u[k].toInt()).toByte()
                if (progress != null && i % PROGRESS_STEP == 0) progress(((block - 1L) * iterations + i).toFloat() / total)
            }
            val n = minOf(hLen, length - offset)
            System.arraycopy(t, 0, out, offset, n)
            offset += n
            block++
        }
        progress?.invoke(1f)
        return out
    }

    internal fun seal(plaintext: ByteArray, key: ByteArray, associatedData: ByteArray): ByteArray {
        val nonce = ByteArray(NONCE_LENGTH).also(random::nextBytes)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_BITS, nonce))
        cipher.updateAAD(associatedData)
        return nonce + cipher.doFinal(plaintext)
    }

    internal fun open(combined: ByteArray, key: ByteArray, associatedData: ByteArray): ByteArray {
        if (combined.size < NONCE_LENGTH + TAG_BITS / 8) throw KeepsakeException.Damaged("encrypted payload")
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_BITS, combined, 0, NONCE_LENGTH))
        cipher.updateAAD(associatedData)
        return try {
            cipher.doFinal(combined, NONCE_LENGTH, combined.size - NONCE_LENGTH)
        } catch (_: AEADBadTagException) {
            // GCM can't tell a wrong key from tampering; a wrong passphrase is by far the likelier.
            throw KeepsakeException.WrongPassphrase()
        } catch (_: GeneralSecurityException) {
            throw KeepsakeException.WrongPassphrase()
        }
    }
}
